File: feature-favorite-venues/src/main/java/home/oleg/feature_favorite_venues/FavoritePlacesFragment.java
package home.oleg.feature_favorite_venues;

import android.content.Context;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v4.app.Fragment;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.support.v7.widget.Toolbar;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import com.smedialink.common.Optional;
import com.smedialink.feature_add_favorite.CreateFavoriteViewModel;

import java.util.ArrayList;
import java.util.List;

import javax.inject.Inject;

import home.oleg.feature_favorite_venues.di.FavoriteVenuesComponent;
import home.oleg.placesnearme.core_presentation.ShowHideBottomBarListener;
import home.oleg.placesnearme.core_presentation.recyclerview.VenueViewItem;
import home.oleg.placesnearme.core_presentation.view_actions.ViewActionObserver;
import home.oleg.placesnearme.core_presentation.viewdata.VenueViewData;

public class FavoritePlacesFragment extends Fragment implements FavoriteVenuesAdapter.FavoriteClicksListener, FavoriteView {

    @Inject
    FavoritePlacesViewModel favoritePlacesViewModel;

    @Inject
    CreateFavoriteViewModel createFavoriteViewModel;

    @Inject
    FavoriteVenuesAdapter adapter;

    private ShowHideBottomBarListener showHideBottomBarListener;

    @Override
    public void onAttach(Context context) {
        injectDependencies();
        super.onAttach(context);
        if (context instanceof ShowHideBottomBarListener) {
            this.showHideBottomBarListener = (ShowHideBottomBarListener) context;
        }
    }

    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        return inflater.inflate(R.layout.fragment_favorite_places, container, false);
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        RecyclerView rvFavorites = view.findViewById(R.id.rvFavoriteVenues);

        rvFavorites.setLayoutManager(new LinearLayoutManager(getContext()));
        rvFavorites.setAdapter(adapter);

        final Toolbar toolbar = view.findViewById(R.id.toolbar);
        Optional.of((AppCompatActivity) getActivity())
                .ifPresent(appCompatActivity -> {
                    appCompatActivity.setSupportActionBar(toolbar);
                    appCompatActivity.getSupportActionBar().setTitle(R.string.fragment_title_favorite);
                });

        favoritePlacesViewModel.getObserver().observe(this, ViewActionObserver.create(this));
    }

    @Override
    public void showData(List<VenueViewItem> detailedVenues) {
        adapter.submitList(new ArrayList<>(detailedVenues));
    }

    @Override
    public void showEmpty() {
        adapter.showEmpty();
    }

    @Override
    public void favoriteClicked(VenueViewData venueViewData) {
        createFavoriteViewModel.manageFavorite(venueViewData);
    }

    @Override
    public void onItemClicked(VenueViewData venueViewData) {
        showHideBottomBarListener.showVenueDetail(venueViewData);
    }

    private void injectDependencies() {
        FavoriteVenuesComponent.Injector.inject(this);
    }

}

File: feature-favorite-venues/src/main/java/home/oleg/feature_favorite_venues/FavoritePlacesFragment.kt
package home.oleg.feature_favorite_venues

import android.content.Context
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.appcompat.app.AppCompatActivity
import androidx.appcompat.widget.Toolbar
import androidx.fragment.app.Fragment
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.smedialink.feature_add_favorite.CreateFavoriteViewModel
import home.oleg.feature_favorite_venues.di.FavoriteVenuesComponent
import home.oleg.placesnearme.core_presentation.ShowHideBottomBarListener
import home.oleg.placesnearme.core_presentation.extensions.observeNonNull
import home.oleg.placesnearme.core_presentation.recyclerview.ItemViewType
import home.oleg.placesnearme.core_presentation.viewdata.VenueViewData
import javax.inject.Inject

class FavoritePlacesFragment : Fragment(), FavoriteVenuesAdapter.FavoriteClicksListener {

    @Inject
    lateinit var favoritePlacesViewModel: FavoritePlacesViewModel

    @Inject
    lateinit var createFavoriteViewModel: CreateFavoriteViewModel

    @Inject
    lateinit var adapter: FavoriteVenuesAdapter

    lateinit var showHideBottomBarListener: ShowHideBottomBarListener

    override fun onAttach(context: Context?) {
        injectDependencies()
        super.onAttach(context)
        //todo -> navigator
        if (context is ShowHideBottomBarListener) {
            this.showHideBottomBarListener = context
        }
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?,
                              savedInstanceState: Bundle?): View? {
        return inflater.inflate(R.layout.fragment_favorite_places, container, false)
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        val rvFavorites = view.findViewById<RecyclerView>(R.id.rvFavoriteVenues)

        rvFavorites.layoutManager = LinearLayoutManager(context)
        rvFavorites.adapter = adapter

        val toolbar = view.findViewById<Toolbar>(R.id.toolbar)
        (activity as AppCompatActivity?)?.apply {
            setSupportActionBar(toolbar)
            supportActionBar?.setTitle(R.string.fragment_title_favorite)
        }

        favoritePlacesViewModel.state.observeNonNull(this) { venueViewItems ->
            if (venueViewItems.isEmpty()) {
                adapter.showEmpty()
            } else {
                adapter.submitList(ArrayList<ItemViewType>(venueViewItems))
            }
        }
    }

    override fun favoriteClicked(venueViewData: VenueViewData) {
        createFavoriteViewModel.manageFavorite(venueViewData)
    }

    override fun onItemClicked(venueViewData: VenueViewData) {
        showHideBottomBarListener.showVenueDetail(venueViewData)
    }

    private fun injectDependencies() {
        FavoriteVenuesComponent.Injector.inject(this)
    }

}
